Adopt more smart pointers in CachedResourceLoader.cpp

cdumez opened the pull request:
#### 9ae716648ccce6cbd2c4b5c6631084fed43e9343
<pre>
Adopt more smart pointers in CachedResourceLoader.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=268141">https://bugs.webkit.org/show_bug.cgi?id=268141</a>

Reviewed by Brent Fulgham.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedTopOrigin const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::checkedFrameLoader const):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::protectedOrigin const):
* Source/WebCore/loader/cache/CachedResource.h:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestImage):
(WebCore::CachedResourceLoader::requestUserCSSStyleSheet):
(WebCore::CachedResourceLoader::checkInsecureContent const):
(WebCore::CachedResourceLoader::protectedFrame const):
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::canRequestAfterRedirection const):
(WebCore::updateRequestFetchMetadataHeaders):
(WebCore::CachedResourceLoader::computeFetchMetadataSite):
(WebCore::CachedResourceLoader::updateRequestAfterRedirection):
(WebCore::CachedResourceLoader::canRequestInContentDispositionAttachmentSandbox const):
(WebCore::CachedResourceLoader::shouldContinueAfterNotifyingLoadedFromMemoryCache):
(WebCore::CachedResourceLoader::shouldUpdateCachedResourceWithCurrentRequest):
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest):
(WebCore::CachedResourceLoader::prepareFetch):
(WebCore::CachedResourceLoader::updateHTTPRequestHeaders):
(WebCore::CachedResourceLoader::protectedDocumentLoader const):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::revalidateResource):
(WebCore::CachedResourceLoader::loadResource):
(WebCore::CachedResourceLoader::determineRevalidationPolicy const):
(WebCore::CachedResourceLoader::printAccessDeniedMessage const):
(WebCore::CachedResourceLoader::cachePolicy const):
(WebCore::CachedResourceLoader::clearDocumentLoader):
(WebCore::CachedResourceLoader::loadDone):
(WebCore::CachedResourceLoader::allCachedSVGImages const):
(WebCore::CachedResourceLoader::preload):
(WebCore::CachedResourceLoader::warnUnusedPreloads):
(WebCore::CachedResourceLoader::isPreloaded const):
(WebCore::CachedResourceLoader::visibleResourcesToPrioritize):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
(WebCore::ResourceCacheValidationSuppressor::ResourceCacheValidationSuppressor):
(WebCore::ResourceCacheValidationSuppressor::~ResourceCacheValidationSuppressor):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::protectedOrigin const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::protectedSettings const):
(WebCore::Page::prioritizeVisibleResources):
(WebCore::Page::protectedCookieJar const):
* Source/WebCore/page/Page.h:

Canonical link: <a href="https://commits.webkit.org/273578@main">https://commits.webkit.org/273578@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f1966507f14323a8cab41bc06e415551f30af2ce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [⏳ 🧪 style ](https://ews-build.webkit.org/#/builders/Style-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios ](https://ews-build.webkit.org/#/builders/iOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 wincairo ](https://ews-build.webkit.org/#/builders/WinCairo-EWS "Waiting in queue, processing has not started yet") 
| [⏳ 🧪 bindings ](https://ews-build.webkit.org/#/builders/Bindings-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 ios-sim ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 wpe-wk2 ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitperl ](https://ews-build.webkit.org/#/builders/WebKitPerl-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2 ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-mac ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-wpe ](https://ews-build.webkit.org/#/builders/WPE-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 webkitpy ](https://ews-build.webkit.org/#/builders/WebKitPy-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 ios-wk2-wpt ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk1 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc ](https://ews-build.webkit.org/#/builders/JSC-Tests-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-ios ](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2 ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 gtk-wk2 ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🛠 🧪 jsc-arm64 ](https://ews-build.webkit.org/#/builders/JSC-Tests-arm64-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv ](https://ews-build.webkit.org/#/builders/tvOS-17-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Sonoma-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 api-gtk ](https://ews-build.webkit.org/#/builders/GTK-Build-EWS "Waiting in queue, processing has not started yet") | 
| [⏳ 🧪 services ](https://ews-build.webkit.org/#/builders/Services-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 tv-sim ](https://ews-build.webkit.org/#/builders/tvOS-17-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🧪 mac-wk2-stress ](https://ews-build.webkit.org/#/builders/macOS-Monterey-Release-Build-EWS "Waiting in queue, processing has not started yet") | [⏳ 🛠 jsc-armv7 ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8170 "Built successfully and passed tests") | [⏳ 🛠 watch ](https://ews-build.webkit.org/#/builders/watchOS-10-Build-EWS "Waiting in queue, processing has not started yet") | | [⏳ 🧪 jsc-armv7-tests ](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Build-EWS "Waiting in queue, processing has not started yet") | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-10-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1052,30 +1062,30 @@ ResourceErrorOr<CachedResourceHandle<CachedResource>> CachedResourceLoader::requ
+                    sameOriginRequest = topDocument->protectedSecurityOrigin()->isSameSchemeHostPort(requestedOrigin.get());
+            } else if (RefPtr document = this->document()) {
+                sameOriginRequest = document->protectedTopOrigin()->isSameSchemeHostPort(requestedOrigin.get())
+                && document->protectedSecurityOrigin()->isSameSchemeHostPort(requestedOrigin.get());
```
nit: It seems like this line should have one more indent

brentfulgham commented on Source/WebCore/loader/cache/CachedResourceLoader.cpp:
```diff
@@ -1052,30 +1062,30 @@ ResourceErrorOr<CachedResourceHandle<CachedResource>> CachedResourceLoader::requ
-                auto* localMainFrame = dynamicDowncast<LocalFrame>(frame.mainFrame());
-                if (frame.isMainFrame())
+                RefPtr localMainFrame = dynamicDowncast<LocalFrame>(frame->mainFrame());
+                if (frame->isMainFrame())
```
The way we use `mainFrame` here (as both the true main frame, and the local "main frame" of an iframe is so confusing. (I know this isn't your new code -- it just is jarring to read).

brentfulgham replied:
Perhaps `mainFrame` should be called `localFrame` or something.

cdumez replied:
This is because of site isolation. The main frame may be local or remote (in another process) once you have out-of-process iframes. It does add some complexity.

brentfulgham commented on Source/WebCore/loader/cache/CachedResourceLoader.h:
```diff
@@ -233,16 +236,16 @@ class ResourceCacheValidationSuppressor {
     ~ResourceCacheValidationSuppressor()
     {
-        m_loader.m_allowStaleResources = m_previousState;
+        m_loader->m_allowStaleResources = m_previousState;
```
Should we consider the possibility that `m_loader` has been cleared at this point, since it's a WeakRef and we don't control it?

cdumez replied:
It's a WeakRef<> (not WeakPtr) which indicates that we don't expect it to become null. As a result, you can / shouldn't null check it. It is does get cleared, then we'll crash safely with a null-deref which is what we want from a security standpoint.

darinadler replied:
can’t / shouldn’t

brentfulgham reviewed (approved):
Looks good. r=me

cdumez committed:
Adopt more smart pointers in CachedResourceLoader.cpp
https://bugs.webkit.org/show_bug.cgi?id=268141

Reviewed by Brent Fulgham.

* Source/WebCore/dom/Document.cpp:
(WebCore::Document::protectedTopOrigin const):
* Source/WebCore/dom/Document.h:
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::checkedFrameLoader const):
* Source/WebCore/loader/DocumentLoader.h:
* Source/WebCore/loader/cache/CachedResource.cpp:
(WebCore::CachedResource::protectedOrigin const):
* Source/WebCore/loader/cache/CachedResource.h:
* Source/WebCore/loader/cache/CachedResourceLoader.cpp:
(WebCore::CachedResourceLoader::requestImage):
(WebCore::CachedResourceLoader::requestUserCSSStyleSheet):
(WebCore::CachedResourceLoader::checkInsecureContent const):
(WebCore::CachedResourceLoader::protectedFrame const):
(WebCore::CachedResourceLoader::canRequest):
(WebCore::CachedResourceLoader::canRequestAfterRedirection const):
(WebCore::updateRequestFetchMetadataHeaders):
(WebCore::CachedResourceLoader::computeFetchMetadataSite):
(WebCore::CachedResourceLoader::updateRequestAfterRedirection):
(WebCore::CachedResourceLoader::canRequestInContentDispositionAttachmentSandbox const):
(WebCore::CachedResourceLoader::shouldContinueAfterNotifyingLoadedFromMemoryCache):
(WebCore::CachedResourceLoader::shouldUpdateCachedResourceWithCurrentRequest):
(WebCore::CachedResourceLoader::updateCachedResourceWithCurrentRequest):
(WebCore::CachedResourceLoader::prepareFetch):
(WebCore::CachedResourceLoader::updateHTTPRequestHeaders):
(WebCore::CachedResourceLoader::protectedDocumentLoader const):
(WebCore::CachedResourceLoader::requestResource):
(WebCore::CachedResourceLoader::revalidateResource):
(WebCore::CachedResourceLoader::loadResource):
(WebCore::CachedResourceLoader::determineRevalidationPolicy const):
(WebCore::CachedResourceLoader::printAccessDeniedMessage const):
(WebCore::CachedResourceLoader::cachePolicy const):
(WebCore::CachedResourceLoader::clearDocumentLoader):
(WebCore::CachedResourceLoader::loadDone):
(WebCore::CachedResourceLoader::allCachedSVGImages const):
(WebCore::CachedResourceLoader::preload):
(WebCore::CachedResourceLoader::warnUnusedPreloads):
(WebCore::CachedResourceLoader::isPreloaded const):
(WebCore::CachedResourceLoader::visibleResourcesToPrioritize):
* Source/WebCore/loader/cache/CachedResourceLoader.h:
(WebCore::ResourceCacheValidationSuppressor::ResourceCacheValidationSuppressor):
(WebCore::ResourceCacheValidationSuppressor::~ResourceCacheValidationSuppressor):
* Source/WebCore/loader/cache/CachedResourceRequest.h:
(WebCore::CachedResourceRequest::protectedOrigin const):
* Source/WebCore/page/Page.cpp:
(WebCore::Page::protectedSettings const):
(WebCore::Page::prioritizeVisibleResources):
(WebCore::Page::protectedCookieJar const):
* Source/WebCore/page/Page.h:

Canonical link: https://commits.webkit.org/273578@main
